### ISSUE TITLE
[scheduler] pod was pending while the node has enough resource

### ISSUE THREAD

**Issue by jianxingzhe** [kubernetes/kubernetes#86942]
**What happened**:
when creating an statefulset with three replicas, only two pods were running, and one pod was pending because of `Insufficient memory`. echo pod requests 4c12g resource,  and has node affinity to node1, node2 and node3. 
```
pt-4c12g-0                           1/1     Running   2          10m     10.197.129.90    node1   <none>           <none>
pt-4c12g-1                           1/1     Running   1          10m     10.197.129.118   node2   <none>           <none>
pt-4c12g-2                           0/1     Pending   0          9m51s   <none>           <none>                   <none>           <none>
```

```
Warning  FailedScheduling  48s (x150 over 11m)  default-scheduler  0/12 nodes are available: 3 Insufficient memory, 9 node(s) didn't match node selector.
```

however, the node3 has enough resources:
```
kubectl  describe node node3
Name:               node3
Roles:              node
Labels:             app=nzk
                    beta.kubernetes.io/arch=amd64
                    beta.kubernetes.io/instance-type=bareMetal
                    beta.kubernetes.io/os=linux
                    dedicated=kafka
                    failure-domain.beta.kubernetes.io/zone=az1
                    kubernetes.io/hostname=node3
                    node-role.kubernetes.io/node=true
Annotations:        node.alpha.kubernetes.io/ttl: 0
                    volumes.kubernetes.io/controller-managed-attach-detach: true
CreationTimestamp:  Tue, 19 Nov 2019 10:06:35 +0800
Taints:             dedicated=kafka:NoSchedule
Unschedulable:      false
Conditions:
  Type                 Status  LastHeartbeatTime                 LastTransitionTime                Reason                       Message
  ----                 ------  -----------------                 ------------------                ------                       -------
  KernelDeadlock       False   Thu, 12 Dec 2019 14:25:30 +0800   Mon, 02 Dec 2019 19:08:47 +0800   KernelHasNoDeadlock          kernel has no deadlock
  ReadonlyFilesystem   False   Thu, 12 Dec 2019 14:25:30 +0800   Mon, 02 Dec 2019 19:08:47 +0800   FilesystemIsNotReadOnly      Filesystem is not read-only
  MemoryPressure       False   Thu, 12 Dec 2019 14:25:39 +0800   Mon, 02 Dec 2019 19:08:41 +0800   KubeletHasSufficientMemory   kubelet has sufficient memory available
  DiskPressure         False   Thu, 12 Dec 2019 14:25:39 +0800   Mon, 02 Dec 2019 19:08:41 +0800   KubeletHasNoDiskPressure     kubelet has no disk pressure
  PIDPressure          False   Thu, 12 Dec 2019 14:25:39 +0800   Mon, 02 Dec 2019 19:08:41 +0800   KubeletHasSufficientPID      kubelet has sufficient PID available
  Ready                True    Thu, 12 Dec 2019 14:25:39 +0800   Thu, 12 Dec 2019 13:44:26 +0800   KubeletReady                 kubelet is posting ready status
Addresses:
  InternalIP:  10.197.128.14
  Hostname:    node3
Capacity:
 cpu:                48
 ephemeral-storage:  1135570792Ki
 hugepages-1Gi:      0
 hugepages-2Mi:      0
 memory:             32314408Ki
 pods:               110
Allocatable:
 cpu:                32
 ephemeral-storage:  1098870632Ki
 hugepages-1Gi:      0
 hugepages-2Mi:      0
 memory:             32314408Ki
 pods:               110
System Info:
 Machine ID:                 b6a66d6c7df24982aa4c73a7276a2ac0
 System UUID:                4c4c4544-0035-3110-8050-b8c04f5a4832
 Boot ID:                    6334d585-e480-4399-8d8a-dcc78556445d
 Kernel Version:             4.19.75
 OS Image:                   Debian GNU/Linux 9 (stretch)
 Operating System:           linux
 Architecture:               amd64
 Container Runtime Version:  docker://18.6.3
 Kubelet Version:            v1.13.12
 Kube-Proxy Version:         v1.13.12
Non-terminated Pods:         (7 in total)
  Namespace                  Name                              CPU Requests  CPU Limits  Memory Requests  Memory Limits  AGE
  ---------                  ----                              ------------  ----------  ---------------  -------------  ---
  kafka-test                 pt-2c4g-0                         2 (6%)        0 (0%)      4Gi (12%)        0 (0%)         147m
  kafka-test                 pt-4c8g-1                         4 (12%)       0 (0%)      8Gi (25%)        0 (0%)         146m
  kube-system                cleanlog-78xdh                    100m (0%)     100m (0%)   40Mi (0%)        40Mi (0%)      7d21h
  kube-system                ds-vpccni-bm-g22t2                0 (0%)        0 (0%)      40Mi (0%)        40Mi (0%)      23d
  kube-system                kube-proxy-lq94n                  100m (0%)     100m (0%)   100Mi (0%)       100Mi (0%)     23d
  kube-system                localstorage-provisioner-g8mhm    200m (0%)     200m (0%)   100Mi (0%)       100Mi (0%)     5d21h
  kube-system                node-problem-detector-4gp42       100m (0%)     100m (0%)   100Mi (0%)       100Mi (0%)     23d
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource           Requests       Limits
  --------           --------       ------
  cpu                6500m (20%)    500m (1%)
  memory             12668Mi (40%)  380Mi (1%)
  ephemeral-storage  0 (0%)         0 (0%)
Events:
  Type    Reason                   Age   From                             Message
  ----    ------                   ----  ----                             -------
  Normal  Starting                 41m   kubelet, prit1-k8s75.jd.163.org  Starting kubelet.
  Normal  NodeHasSufficientMemory  41m   kubelet, prit1-k8s75.jd.163.org  Node node3 status is now: NodeHasSufficientMemory
  Normal  NodeHasNoDiskPressure    41m   kubelet, prit1-k8s75.jd.163.org  Node node3 status is now: NodeHasNoDiskPressure
  Normal  NodeHasSufficientPID     41m   kubelet, prit1-k8s75.jd.163.org  Node node3 status is now: NodeHasSufficientPID
  Normal  NodeNotReady             41m   kubelet, prit1-k8s75.jd.163.org  Node node3 status is now: NodeNotReady
  Normal  NodeAllocatableEnforced  41m   kubelet, prit1-k8s75.jd.163.org  Updated Node Allocatable limit across pods
  Normal  NodeReady                41m   kubelet, prit1-k8s75.jd.163.org  Node node3 status is now: NodeReady

```

**What you expected to happen**:
the pod shoule be scheduled successfully

**How to reproduce it (as minimally and precisely as possible)**:

the problem will appear with some conditions：
```
1. we have extend the default scheduler's predicate interface
2. the pod use local pv
3. the local pvs are created by localstorage-provisioner dynamically 
```

**Anything else we need to know?**:

the reason of this problem is that the `nominatedPodMap` of scheduler is inconsistent:
```
1. assume podA is scheduled to node1, when podA is waiting for the local pv to be bound in bind phase, the scheduler may try to schedule podA again, as the bind phase is asynchronous.
2. podA pass the default predicates，but failed with extend scheduler predicate interface, then it will enter the preempt phase.
3. in preempt phase, the scheduler does not call extend scheduler predicate interface, so it find a nominated node(node3), then it will update the `nominatedPodMap`.
4. when local pv is bound, the podA will run on node1, but the nominatedPodMap still recorde that podA.Status.nominatedNodeName is node3.
5. when podB is scheduled to node3, node3 may have insufficient memory when minus PodA's resouce.
```
Is that right?

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.12", GitCommit:"a8b52209ee172232b6db7a6e0ce2adc77458829f", GitTreeState:"clean", BuildDate:"2019-10-15T12:12:15Z", GoVersion:"go1.11.13", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.12", GitCommit:"a8b52209ee172232b6db7a6e0ce2adc77458829f", GitTreeState:"clean", BuildDate:"2019-10-15T12:04:30Z", GoVersion:"go1.11.13", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
```
PRETTY_NAME="Debian GNU/Linux 9 (stretch)"
NAME="Debian GNU/Linux"
VERSION_ID="9"
VERSION="9 (stretch)"
VERSION_CODENAME=stretch
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:

**Comment by liggitt** [comment]
duplicate of https://github.com/kubernetes/kubernetes/issues/86626?

**Comment by jianxingzhe** [comment]
> duplicate of #86626?

**Comment by jianxingzhe** [comment]
> duplicate of #86626?

mybe not， this problem occurs during preemption.  when predicating nodes for a pod during preemption, default scheduler doesn't call the extender scheudler interface, which may caused the 
nominatedPodMap cache inconsistency.

**Comment by snowplayfire** [comment]
@jianxingzhe  Can the problem be reproduced stably if I follow your three conditions?

**Comment by jianxingzhe** [comment]
> @jianxingzhe Can the problem be reproduced stably if I follow your three conditions?

in our cluster, it can be reproduced stably. try to create/delete sts for serveral times, it will occur.

**Comment by alculquicondor** [comment]
What is the memory request of your Pod?

**Comment by jianxingzhe** [comment]
> What is the memory request of your Pod?

the memory request of my pod is 12Gi

**Comment by attlee-wang** [comment]
@jianxingzhe  I have a similar problem at [86626](https://github.com/kubernetes/kubernetes/issues/86626), and it is difficult to reproduce it. Since you can reproduce it, you can add some log debugging and help us solve this problem.

**Comment by alculquicondor** [comment]
Could you try to repro in 1.16 or above? We have migrated a lot of checks to internal caches so it's possible that this problem is solved.
Also, getting fixes for 1.13 is pretty much impossible, as we only maintain the last 3 releases.

**Comment by JasonRD** [comment]
Having the same problem with scheduler predicate extender, like @jianxingzhe describe above, the root cause is inconsistent nominatedPodMap. In one scheduling cycle, pod will be process in the sequence of predicate, preempt, assumed, bind. 
My scenario is: 

1. In the scheduling cycle A, podA is being scheduling, It is changed by scheduler extender, such as update pod annotations, etc. 
2. The Kubernetes Scheduler get update event before podA being assumed to nodeA in scheduling cycle A, so the function  `skipPodUpdate` will return false. PodA is re-enqueue to activeQ, and will be scheduling in scheduling cycle B, which next to scheduling cycle A.
3. Then, podA  is going to be asynchronously bound with nodeA.
4. In the scheduling cycle B, podA is been popped out of activeQ, and then predicated failed with some reason, such as resource insufficient, or scheduler extender predicate failed (it is  predicated failed by scheduler extender in my scenario). 
5. Then, podA step into preempt phase. It selects some victim pod, and add podA into nominatedPodMap.  
6. At last scheduling cycle B be end, but podA will always stay alone in the nominatedPodMap, even if podA had been deleted.

Because in the predicate phase, function `podFitsOnNode` will always check resources with added nominated pod resources by function `addNominatedPods`, it cause the pod being pending but with enough resources.

**Comment by alculquicondor** [comment]
@JasonRD what version are you running?

**Comment by JasonRD** [comment]
@alculquicondor we are running version 1.14 on production environment, and have test version 1.17.7, 1.18.2, which also have the problem.

**Comment by alculquicondor** [comment]
I don't have time to work on this, but here are my thoughts.

We call `skipPodUpdate` at 2 points:

1. When the update is received and we try to update our cache.
1. When we just initiated another scheduling cycle. At this point, we know if a pod was assumed or not, but we don't know if binding will succeed. Is there a reason it wouldn't other than apiserver having an unexpected error?

So, in 2, we have to try to schedule the pod. At this point, two things can happen: (a) the pod still fits, so we will try to bind it again, and this would fail; (b) the pod doesn't fit and we try to do a preemption, and thus it will enter the nominated pod map.

Possible bugs:

1. We don't call extenders during preemption. This we should definitely fix.
1. We don't clear up the nominated pod map for the pod (it could be in a different node than the incoming update suggests), once the first binding succeeds (or during its corresponding pod update).

Alternatively, should we just not try to schedule while a pod is assumed? And fail the scheduling cycle so that the pod enters the backoff?

cc @Huang-Wei who has been looking into pod update events recently.

Also, note that we can only fix versions 1.16+ if we hurry, otherwise 1.17+ after 1.19 has been released.

**Comment by alculquicondor** [comment]
Actually, annotations are skipped from updates. So this might be related to the race conditions that @Huang-Wei was working on for 1.19

**Comment by JasonRD** [comment]
@alculquicondor We checkout the latest code of scheduler, find it had resolved the problem. And it is fixed with [this pr ](https://github.com/kubernetes/kubernetes/pull/86230).

**Comment by alculquicondor** [comment]
#86230 is included in 1.18 though. Could you recheck the latest release of 1.18?

**Comment by Huang-Wei** [comment]
@jianxingzhe As it's impossible to back port fixes to 1.13 (the version you're using), I'd suggest to upgrade to 1.18 to see if the problem still exists.

BTW: In recent releases, we delivered quite a number of fixes on saving unnecessary scheduling attempts. Also, the preemption logic, esp. the nominatedNodeMap parts got revisited and a few fixes have been delivered (in 1.19).

---

Re:

> 3. in preempt phase, the scheduler does not call extend scheduler predicate interface, so it finds a nominated node(node3), then it will update the `nominatedPodMap`.

It's true that the filtering (predictes) part inside preemption doesn't call filter (predicate) extenders. If you think this is the key of this issue, we can definitely provide a fix and backport to supported k8s versions.
